bitcoin/test/fuzz/miniscript_random.cpp

```diff
@@ -309,10 +310,111 @@ std::optional<NodeInfo> ConsumeNode(FuzzedDataProvider& provider) {
     return {};
 }
 
+/**
+ * Consume a Miniscript node from the fuzzer's output.
+ *
+ * This is similar to ConsumeNodeStable, but uses a miniscript::Type-driven heuristics
+ * to construct the nodes. It is intended to more quickly explore interesting miniscripts,
+ * at the cost of higher implementation complexity (which could cause it miss things if
+ * incorrect), and with less regard for stability of the seeds (as changes to the
+ * heuristics implemented here may dramatically alter the resulting node for a given seed).
+ */
+std::optional<NodeInfo> ConsumeNodeSmart(FuzzedDataProvider& provider, miniscript::Type type_needed) {
+    // If no bytes remain in provider, abort. This prevents infinite loops
+    // in case a read zero (for fragcode) results in a recursive rule triggering.
+    if (provider.remaining_bytes() == 0) return {};
+
+    // Not all type properties are implemented in the match logic below,
+    // so strip away the ones we cannot discern. When the node is actually
+    // constructed, we compare the full requested type properties.
+    auto typ = type_needed & "BVWKzondu"_mst;
+    // Some helpers for computing the child nodes' type properties.
+    auto base = "BVK"_mst & typ;
+    auto tz = typ & "z"_mst;
+    auto toz = "z"_mst.If(typ << "o"_mst);
+    auto to = typ & "o"_mst;
+    auto tn = typ & "n"_mst;
+    auto td = typ & "d"_mst;
+    auto tu = typ & "u"_mst;
+
+    // Fragcode selects which of the (applicable) matching rules below is selected.
+    // Every rule, if it matches, checks if fragcode has reached 0, and if so,
+    // the rule is used. If not, fragcode is decremented and we continue to the
+    // next rule. This is performed in a loop so that if all rules were tried,
+    // and fragcode hasn't reached 0 yet, we start over. This avoids the need to
+    // count the number of matching rules up front.
+    int fragcode = provider.ConsumeIntegral<uint8_t>();
+
+    while (true) {
+        /** Counter for the number of applicable rules. */
+        int candidates = 0;
+
+        // Rules for constructing leaf nodes
+        if ("Bzud"_mst << typ && ++candidates && !(fragcode--)) return {{Fragment::JUST_0}};
```

Comment:
The way to think about it is that types are sets of potential values a variable can take. A subtype is called so because it corresponds to a subset of the potential values. Since the empty type here corresponds to every possible valid expression, every other type is indeed a subset of it.